numpyro/util.py

```diff
@@ -331,48 +332,77 @@ def _body_fn(i, vals):
         val, collection, start_idx, thinning = vals
         val = body_fun(val)
         idx = (i - start_idx) // thinning
-        collection = cond(
-            idx >= 0,
-            collection,
-            lambda x: x.at[idx].set(ravel_pytree(transform(val))[0]),
-            collection,
-            identity,
-        )
+
+        def update_fn(collect_array, new_val):
+            return cond(
+                idx >= 0,
+                collect_array,
+                lambda x: x.at[idx].set(new_val),
+                collect_array,
+                identity,
+            )
+
+        def update_collection(collection, val):
+            return jax.tree.map(update_fn, collection, transform(val))
+
+        collection = jit(update_collection, donate_argnums=0)(collection, val)
+
         return val, collection, start_idx, thinning
 
-    collection = jnp.zeros(
-        (collection_size,) + init_val_flat.shape, dtype=init_val_flat.dtype
-    )
+    @partial(jit, donate_argnums=2)
+    @cached_by(fori_collect, body_fun, transform)
+    def _body_fn_wrap(i, val, collection, start_idx, thinning):
+        return _body_fn(i, (val, collection, start_idx, thinning))
+
+    def map_fn(x):
+        nx = jnp.asarray(x)
+        return jnp.zeros((collection_size, *nx.shape), dtype=nx.dtype) * nx[None, ...]
+
+    collection = jax.tree.map(map_fn, init_val_transformed)
+
     if not progbar:
-        last_val, collection, _, _ = fori_loop(
-            0, upper, _body_fn, (init_val, collection, start_idx, thinning)
-        )
+
+        def loop_fn(collection):
+            return fori_loop(
+                0, upper, _body_fn, (init_val, collection, start_idx, thinning)
+            )
+
+        last_val, collection, _, _ = jit(loop_fn, donate_argnums=0)(collection)
     elif num_chains > 1:
         progress_bar_fori_loop = progress_bar_factory(upper, num_chains)
         _body_fn_pbar = progress_bar_fori_loop(_body_fn)
-        last_val, collection, _, _ = fori_loop(
-            0, upper, _body_fn_pbar, (init_val, collection, start_idx, thinning)
-        )
+
+        def loop_fn(collection):
+            return fori_loop(
+                0, upper, _body_fn_pbar, (init_val, collection, start_idx, thinning)
+            )
+
+        last_val, collection, _, _ = jit(loop_fn, donate_argnums=0)(collection)
+
     else:
         diagnostics_fn = progbar_opts.pop("diagnostics_fn", None)
         progbar_desc = progbar_opts.pop("progbar_desc", lambda x: "")
 
         vals = (init_val, collection, device_put(start_idx), device_put(thinning))
+        # jitted_body_fn_wrap = jit(_body_fn_wrap)
+
         if upper == 0:
             # special case, only compiling
-            jit(_body_fn)(0, vals)
+            val, collection, start_idx, thinning = vals
+            dummy_collection = jax.tree.map(lambda x: x.copy(), collection)
```

Comment:
If collection was used then it would be donated or modified and couldn't be used later.  I swapped the `body_fn_wrap` call to use a -1 index so now collection can be used and remain unmodified.